// @hiveio/hive-js/lib/broadcast/index.js
'use strict';

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var _debug = require('debug');

var _debug2 = _interopRequireDefault(_debug);

var _helpers = require('./helpers');

var _helpers2 = _interopRequireDefault(_helpers);

var _formatter = require('../formatter');

var _formatter2 = _interopRequireDefault(_formatter);

var _api = require('../api');

var _api2 = _interopRequireDefault(_api);

var _auth = require('../auth');

var _auth2 = _interopRequireDefault(_auth);

var _utils = require('../utils');

var _operations = require('../auth/serializer/src/operations');

var _ecc = require('../auth/ecc');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var operations = require('./operations');
var config = require('../config');

var debug = (0, _debug2.default)('hive:broadcast');
var noop = function noop() {};
var formatter = (0, _formatter2.default)(_api2.default);

function getTransactionStatus(trxId, expiration) {
  var time = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 3000;

  return new _bluebird2.default(function (resolve, reject) {
    setTimeout(function () {
      _api2.default.findTransactionAsync(trxId, expiration).then(function (res) {
        resolve(res);
      });
    }, time);
  });
}

var hiveBroadcast = {};

// Base transaction logic -----------------------------------------------------

/**
 * Sign and broadcast transactions on the hive network
 */

hiveBroadcast.send = function hiveBroadcast$send(tx, privKeys, callback) {
  var trxId = void 0;
  var resultP = hiveBroadcast._prepareTransaction(tx).then(function (transaction) {
    if (config.get("address_prefix") === "TST") {
      transaction.operations = JSON.parse(JSON.stringify(transaction.operations).replace(/ HIVE/g, ' TESTS'));
      transaction.operations = JSON.parse(JSON.stringify(transaction.operations).replace(/ HBD/g, ' TBD'));
    }
    debug('Signing transaction (transaction, transaction.operations)', transaction, transaction.operations);
    var buf = _operations.transaction.toBuffer(transaction);
    trxId = _ecc.hash.sha256(buf).toString('hex').slice(0, 40);
    return _bluebird2.default.join(transaction, _auth2.default.signTransaction(transaction, privKeys));
  }).spread(function (transaction, signedTransaction) {
    debug('Broadcasting transaction (transaction, transaction.operations)', transaction, transaction.operations);
    return _api2.default.broadcastTransactionAsync(signedTransaction).then(function (result) {
      return Object.assign({ id: trxId }, result, signedTransaction);
    });
  });

  resultP.nodeify(callback || noop);
};

hiveBroadcast._prepareTransaction = function hiveBroadcast$_prepareTransaction(tx) {
  var propertiesP = _api2.default.getDynamicGlobalPropertiesAsync();
  return propertiesP.then(function (properties) {
    // Set defaults on the transaction
    var chainDate = new Date(properties.time + 'Z');
    var refBlockNum = properties.last_irreversible_block_num - 1 & 0xFFFF;
    return _api2.default.getBlockHeaderAsync(properties.last_irreversible_block_num).then(function (block) {
      var headBlockId = block ? block.previous : '0000000000000000000000000000000000000000';
      return Object.assign({
        ref_block_num: refBlockNum,
        ref_block_prefix: new Buffer(headBlockId, 'hex').readUInt32LE(4),
        expiration: new Date(chainDate.getTime() + 600 * 1000)
      }, tx);
    });
  });
};

// Generated wrapper ----------------------------------------------------------
operations.forEach(function (operation) {
  var operationName = (0, _utils.camelCase)(operation.operation);
  var operationParams = operation.params || [];

  var useCommentPermlink = operationParams.indexOf('parent_author') !== -1 && operationParams.indexOf('parent_permlink') !== -1;

  hiveBroadcast[operationName + 'With'] = function hiveBroadcast$specializedSendWith(wif, options, callback) {
    debug('Sending operation "' + operationName + '" with', { options: options, callback: callback });
    var keys = {};
    if (operation.roles && operation.roles.length) {
      keys[operation.roles[0]] = wif; // TODO - Automatically pick a role? Send all?
    }

    return hiveBroadcast.send({
      extensions: [],
      operations: [[operation.operation, Object.assign({}, options, options.json_metadata != null ? {
        json_metadata: toString(options.json_metadata)
      } : {}, useCommentPermlink && options.permlink == null ? {
        permlink: formatter.commentPermlink(options.parent_author, options.parent_permlink)
      } : {})]]
    }, keys, callback);
  };

  hiveBroadcast[operationName] = function hiveBroadcast$specializedSend(wif) {
    for (var _len = arguments.length, args = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
      args[_key - 1] = arguments[_key];
    }

    debug('Parsing operation "' + operationName + '" with', { args: args });
    var options = operationParams.reduce(function (memo, param, i) {
      memo[param] = args[i]; // eslint-disable-line no-param-reassign
      return memo;
    }, {});
    var callback = args[operationParams.length];
    return hiveBroadcast[operationName + 'With'](wif, options, callback);
  };
});

hiveBroadcast.updateOperations = function () {
  console.log('Warning: call to updateOperations() is deprecated and can safely be removed');
};

var toString = function toString(obj) {
  return (typeof obj === 'undefined' ? 'undefined' : _typeof(obj)) === 'object' ? JSON.stringify(obj) : obj;
};
(0, _helpers2.default)(hiveBroadcast);

_bluebird2.default.promisifyAll(hiveBroadcast);

exports = module.exports = hiveBroadcast;
